### PR TITLE
Add immutable withTheme method

### PR DESCRIPTION
**Pull request by Gerych1984**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: src/ViewTrait.php [hunk 1]
```diff
@@ -251,6 +250,14 @@ public function setTheme(?Theme $theme): static
         return $this;
     }
 
+    public function withTheme(?Theme $theme): static
+    {
+        $new = clone $this;
+        $new->state = clone $this->state;
```
Comment (vjik) [inline src/ViewTrait.php:256]: State do not should clone, parameteres and blocks from state should keep. 

I think need extract theme to separate `ThemeState`  by analogy with `LocaleState`.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/ViewTrait.php (hunk 1 above)
Comment (Gerych1984) [reply src/ViewTrait.php]: > State should not be cloned, parameters and blocks from the state should be kept.
> 
> I think need extract theme to separate `ThemeState` by analogy with `LocaleState`.

Done

File: src/State/ThemeState.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    public function __construct(
+        private ?Theme $theme = null
+    ) {
+    }
+
+    /**
+     * Set the specified view theme.
+     *
+     * @param Theme|null $theme
+     * @return $this
```
Comment (vjik) [inline src/State/ThemeState.php:23]:
```suggestion
     * @param Theme|null $theme The theme instance or `null` for reset theme.
```

File: src/State/ThemeState.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $this->theme = $theme;
+
+        return $this;
+    }
+
+    /**
+     * Gets the current theme or null if not set
+     *
+     * @return Theme|null
```
Comment (vjik) [inline src/State/ThemeState.php:35]:
```suggestion
     * Gets the theme instance, or `null` if no theme has been set.
     *
     * @return Theme|null The theme instance, or `null` if no theme has been set.
```

File: src/View.php [hunk 4]
```diff
@@ -12,8 +12,8 @@
 use Yiisoft\View\Event\View\PageBegin;
 use Yiisoft\View\Event\View\PageEnd;
 use Yiisoft\View\State\LocaleState;
+use Yiisoft\View\State\ThemeState;
 use Yiisoft\View\State\ViewState;
-
 use function ob_end_flush;
```
Comment (vjik) [inline src/View.php:17]:
```suggestion

use function ob_end_flush;
```

File: src/WebView.php [hunk 5]
```diff
@@ -21,8 +21,8 @@
 use Yiisoft\View\Event\WebView\PageBegin;
 use Yiisoft\View\Event\WebView\PageEnd;
 use Yiisoft\View\State\LocaleState;
+use Yiisoft\View\State\ThemeState;
 use Yiisoft\View\State\WebViewState;
-
 use function array_merge;
```
Comment (vjik) [inline src/WebView.php:26]:
```suggestion

use function array_merge;
```

**Review by vjik — CHANGES_REQUESTED** [review]
Also need:

- Add line to changelog
- Clear theme state in `ViewTrait::clear()`

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 ## 8.0.1 under development
 
+- Chg: Add immutable `ViewTrait::withTheme()` method and new `ThemeState` class
```
Comment (vjik) [inline CHANGELOG.md:5]:
```suggestion
- New #243: Add immutable method `ViewInterface::withTheme()` (@Gerych1984)
```

File: src/State/StateTrait.php [hunk 7]
```diff
@@ -5,8 +5,6 @@
 namespace Yiisoft\View\State;
 
 use InvalidArgumentException;
-use Yiisoft\View\Theme;
-
 use function array_key_exists;
```
Comment (vjik) [inline src/State/StateTrait.php:8]:
```suggestion

use function array_key_exists;
```

File: src/State/ThemeState.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    public function __construct(
+        private ?Theme $theme = null
+    ) {
+    }
+
+    /**
+     * Set the specified view theme.
+     *
+     * @param Theme|null $theme $theme The theme instance or `null` for reset theme.
+     * @return $this
```
Comment (vjik) [inline src/State/ThemeState.php:23]:
```suggestion
```

File: tests/ViewTest.php [hunk 9]
```diff
@@ -19,7 +19,6 @@
 use Yiisoft\View\Theme;
 use Yiisoft\View\View;
 use Yiisoft\View\ViewContextInterface;
-
 use function crc32;
```
Comment (vjik) [inline tests/ViewTest.php:22]:
```suggestion

use function crc32;
```

File: tests/ViewTest.php [hunk 10]
```diff
@@ -584,6 +583,18 @@ public function testImmutability(): void
         $this->assertNotSame($view, $view->withClearedState());
         $this->assertNotSame($view, $view->withLocale('es'));
         $this->assertNotSame($view, $view->withFallbackExtension('tpl'));
+        $this->assertNotSame($view, $view->withTheme(new Theme([])));
```
Comment (vjik) [inline tests/ViewTest.php:586]:
```suggestion
        $this->assertNotSame($view, $view->withTheme(null));
```

**Review by vjik — APPROVED** [review]

**Comment by vjik** [comment]
👍

**Pull request merged**